### DISCUSSION TITLE
MongoDB integration example

### DISCUSSION THREAD

**Discussion by ch4nd4n** [remix-run/remix#1104] [Q&A] [answered]
Is there any example of how to integrate with MongoDB? I would like to know the best practice around it, any comments?

This is what I did
1. Create Mongoose Model 
```
const TodoSchema = new mongoose.Schema({
  title: { type: String, required: true },
});
const model = mongoose.models.Todo || mongoose.model("Todo", TodoSchema);
```
2. Connect to Mongo

```
mongoose.connect(
  "mongodb://root:example@localhost:27017/chandan_vercel_dev?authSource=admin",
);
```
3. Fetch data using loader

```
export let loader: any = async () => {
  return await Todo.find({});
};
```

Described the workflow at https://www.chandankumar.com/blog/remix-run-mongodb-tutorial

Working example at https://github.com/ch4nd4n/practice/tree/master/remix/firebase-auth

**Comment by Adbib** [comment]
ty to use Prisma ORM its can help you with MongoDB and make your work soo easy

**Comment by ch4nd4n** [comment]
Yes I am aware of that. Wanted to know without prisma.

**Reply by Adbib** [reply]
i will try it sooner and send to you my results

**Reply by ch4nd4n** [reply]
Thanks, I got a working example at https://github.com/ch4nd4n/practice/tree/master/remix/firebase-auth

**Reply by ccssmnn** [reply]
>  I would like to know the best practice around it, any comments?

Best practice is to use Prisma for your databases, if available.

**Reply by ch4nd4n** [reply]
> Best practice is to use Prisma for your databases, if available.

I suspect that's the case on the [official documentation](https://www.prisma.io/docs/concepts/database-connectors/mongodb#known-limitations)

> Known limitations
> As MongoDB support is currently in preview there are some known limitations that come with adopting Prisma into your MongoDB project, these include:

> Currently no embedded collection support.

Embedded documents and the ability to query is one of the core features. I am not sure without these anyone would take Prisma seriously for MongoDB.

**Reply by ccssmnn** [reply]
Yeah, if Prisma support is not there (yet), I would refer to the database's best practices for NodeJS clients. Data fetching inside the loaders is just NodeJS communicating with the client. Everything else is cient-server-communication, which should not be affected by your choice of database.

**Comment by ghost** [comment, marked as answer]
There is nothing special about using MongoDB.
You can create a global connection like in the NextJS [example](https://github.com/vercel/next.js/blob/canary/examples/with-mongodb/lib/mongodb.js) and then use that in your actions/loaders/...
Same for mongoose.

**Reply by ch4nd4n** [reply]
For some reason, I missed the example. Yes, I ended up doing something similar but abstracted the implementation in a service layer.
